src/Dictionaries/CacheDictionary.cpp

```diff
@@ -853,17 +877,53 @@ void CacheDictionary::tryPushToUpdateQueueOrThrow(UpdateUnitPtr & update_unit_pt
                 std::to_string(update_queue.size()));
 }
 
+
+std::vector<CacheDictionary::AttributeValue> CacheDictionary::getAttributeValuesFromBlockAtPosition(const std::vector<const IColumn *> & column_ptrs, size_t position)
+{
+    std::vector<AttributeValue> answer;
+    answer.reserve(column_ptrs.size());
+
+    for (const auto * pure_column : column_ptrs)
+    {
+#define DISPATCH(TYPE) \
+        if (const auto * column = typeid_cast<const Column##TYPE *>(pure_column)) { \
+            answer.emplace_back(column->getElement(position)); \
+            continue; \
+        }
+        DISPATCH(UInt8)
+        DISPATCH(UInt16)
+        DISPATCH(UInt32)
+        DISPATCH(UInt64)
+        DISPATCH(UInt128)
+        DISPATCH(Int8)
+        DISPATCH(Int16)
+        DISPATCH(Int32)
+        DISPATCH(Int64)
+        DISPATCH(Decimal<Decimal32>)
```

Comment:
In createAttribute you have just Decimal32, but here Decimal<Decimal32>, is this intended?
Normally a dispatch macro would take another macro as an argument, and apply it to a list of types. This means you don't have to list the types twice. You could probably also use the same dispatch macro for declaring the AttributeValue variant type. See for example the `APPLY_FOR_AGGREGATED_VARIANTS` macro in Aggregator.

Reply:
This is because in createAttribute I have no columns. Here I try to cast the input column to all others. And we have no ColumnDecimal32, but have ColumnDecimal<Decimal32>.

Reply:
Oh, I see, you're concatenating it to get the type of the column object. It would be good if we had some mapping like C++ data type -> column type, but apparently we don't. If we had it, we could just list the supported attribute types only once, name it e.g. `CACHE_DICTIONARY_APPLY_FOR_ATTRIBUTE_TYPES`, and then you'd write here `CACHE_DICT......TYPES(DISPATCH)`. Although `DISPATCH` is a bit of misnomer, it works with single type and doesn't choose between types.

There are several other places where the list is just the same, but still written out many times. Maybe we can fix at least them?

Reply:
Will think about.